### PR TITLE
Upload artifact take2

### PR DESCRIPTION
**Pull request by SimplyAhmazing**
This PR,

* sets up `yargs` for CLI
* minor refactors for ping command
* adds `init` command which takes a cli arg for bucket `--bucket sample-bucket`
* adds `prep` command which will write `commit`, `branch`, and `platform` vars into a `electron-builder.env` file
* adds tests for commands

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,16 +5,19 @@ os:
 language: node_js
 
 node_js:
-  - '8'
+  - '6'
```
Comment (levthedev) [inline .travis.yml]: I think we should stick to 8 unless it breaks builds. Don't see any reason not to.

File: .travis.yml [hunk 2]
```diff
@@ -5,16 +5,19 @@ os:
...
 
 before_install:
   - npm install -g npm@5.2.0
+  - rm -rf node_modules
```
Comment (levthedev) [inline .travis.yml]: Don't think we actually need this, seems like it was a temporary attempt to fix problems introduced by not installing @node/typings

File: .travis.yml [hunk 3]
```diff
@@ -5,16 +5,19 @@ os:
...
+  - npm i --upgrade -g typescript
   - npm install
  
-cache:
-  directories:
-  - node_modules
+# cache:
+#   directories:
+#   - node_modules
```
Comment (levthedev) [inline .travis.yml:17]: I'd like to turn caching on in the future, but for now it seems to sometimes break things

File: appveyor.yml [hunk 4]
```diff
@@ -8,15 +8,18 @@ os: Previous Visual Studio 2015
 #  - node_modules
 
 install:
-  - ps: Install-Product node 8 x64
+  - ps: Install-Product node 6 x64
```
Comment (levthedev) [inline appveyor.yml]: See above comment about Node 8

File: appveyor.yml [hunk 5]
```diff
@@ -8,15 +8,18 @@ os: Previous Visual Studio 2015
...
   # - git reset --hard HEAD
   - node --version
   - npm --version
   - npm install -g npm@5.2.0
+  - rm -rf node_modules
```
Comment (levthedev) [inline appveyor.yml]: See above comment about this

File: dist/cli.js [hunk 6]
```diff
@@ -1,5 +1,35 @@
...
+    return yarg.demand('bucket');
+})
+    .example('fission --bucket my-s3-bucket', 'Initialize package.json with a bucket')
+    .example('fission init --path custom/path/to/package.json', 'Specify path to package.json')
+    .command('prep', 'Prepare dir for electron-builder publish. Run before starting electron-builder')
+    .example('fission prep', 'Configures electron-builder to upload to S3 bucket')
+    .help('h')
+    .alias('h', 'help')
+    .argv;
+console.log(argv);
```
Comment (levthedev) [inline dist/cli.js]: Probably should remove this

File: src/cli.ts [hunk 7]
```diff
@@ -1,5 +1,41 @@
...
+  ) 
+  .example('fission --bucket my-s3-bucket', 'Initialize package.json with a bucket')
+  .example('fission init --path custom/path/to/package.json', 'Specify path to package.json')
+  .command('prep', 'Prepare dir for electron-builder publish. Run before starting electron-builder')
+  .example('fission prep', 'Configures electron-builder to upload to S3 bucket')
+  .help('h')
+  .alias('h', 'help')
+  .argv
+
+console.log(argv)
```
Comment (levthedev) [inline src/cli.ts]: See above comment, was made on .js file by accident

File: src/init.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import * as fs from 'fs'
+import * as detectIndent from 'detect-indent'
+
+import * as fission from './index'
+
+export function writePublishSection (packageJsonPath : string, publishInfo : Object) {
+  const file = fs.readFileSync(packageJsonPath).toString()
+  const indent = detectIndent(file).indent 
+  const packageJson = JSON.parse(file)
+  packageJson.build.publish = publishInfo
```
Comment (levthedev) [inline src/init.ts:12]: will this work if packageJson.build is undefined?

File: src/prep.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import { spawn } from 'child_process'
+import * as fs from 'fs'
+import * as detectIndent from 'detect-indent'
+
+import * as fission from './index'
+import * as util from './util'
+
+export function getEnvVars () : Map <string, string> {
```
Comment (levthedev) [inline src/prep.ts:10]: Why the use of Map here and not Object or Tuple? Is there some TS advantage?

File: src/prep.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  map.set('COMMIT', util.getCommit())
+  map.set('BRANCH', util.getBranch())
+  map.set('PLATFORM', util.getPlatform())
+  return map
+}
+
+export function writeElectronBuilderEnvFile (path : string, envVars : Map <string, string>) {
+  envVars.forEach((val, key) => {
+    // console.log('key', key, 'val', val)
+    fs.appendFileSync(path, `${key}=${val}\n`)
```
Comment (levthedev) [inline src/prep.ts]: Probably better to construct whole string with loop, then append once to file, as opposed to looping through vars and appending for each.

Seems more error prone to write multiple times, no?

File: src/prep.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  })
+}
+
+export function prep (envFilePath : string) {
+  writeElectronBuilderEnvFile(envFilePath, getEnvVars())
```
Comment (levthedev) [inline src/prep.ts]: Why the wrapped export? Can't whoever calls this method just call writeElectronBuilderenvFile with correct args?

If so, maybe rename writeElectronBuilderEnvFile to Ping and just have the caller provide args

File: src/util.ts [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    throw new Error('package.json:publish.provider key must have value "s3"')
+  } else if (!s3PublishInfo.bucket) {
+    throw new Error('package.json:publish.bucket key must have a bucket name')
+  } else if (s3PublishInfo.path !== '${env.BRANCH}/${env.COMMIT}/${env.PLATFORM}') {
+    throw new Error('package.json:publish.path key must be set to "${env.BRANCH}/${env.COMMIT}/${env.PLATFORM}"')
+  }
+}
+
+export function getAppPlatform(): string {
+  if (process.env.APP_PLATFORM) {
```
Comment (levthedev) [inline src/util.ts:19]: Why the conditional? Why not just return `process.env.APP_PLATFORM || process.env.PLATFORM || process.env.TRAVIS_OS_NAME || os.platform()`

File: src/util.ts [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    return process.env.APP_PLATFORM || 'N/A'
+  }
+  return process.env.PLATFORM || process.env.TRAVIS_OS_NAME || os.platform()
+}
+
+export function isTravis(): boolean {
+  return process.env.TRAVIS ? true : false
+}
+
+export function isAppVeyor() {
```
Comment (levthedev) [inline src/util.ts]: Should have boolean type sig

File: tsconfig.json [hunk 14]
```diff
@@ -3,12 +3,18 @@
     "lib": ["es2015"],
     "target": "es5",
     "module": "commonjs",
-    "declaration": true,
+    "declaration": false,
```
Comment (levthedev) [inline tsconfig.json:6]: Why the change?

File: tsconfig.json [hunk 15]
```diff
@@ -3,12 +3,18 @@
...
     "outDir": "./dist",
+    "allowJs": true,
```
Comment (levthedev) [inline tsconfig.json:8]: Why the change?

File: tsconfig.json [hunk 16]
```diff
@@ -3,12 +3,18 @@
...
     "strict": true
   },
   "include": [
     "src/*.ts",
+    "src/*.js",
```
Comment (levthedev) [inline tsconfig.json:13]: See above

File: test/prep.spec.ts [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    }
+    const expectedEnvFilePath = path.join(__dirname, 'data', 'electron-builder.env')
+
+    const envVars = new Map([["COMMIT", "foo"], ["BRANCH", "bar"]])
+    writeElectronBuilderEnvFile(emptyEnvFilePath, envVars)
+
+    const result = fs.readFileSync(emptyEnvFilePath)
+    const expected = fs.readFileSync(expectedEnvFilePath)
+    console.log(result.toString())
+    console.log(expected.toString())
```
Comment (levthedev) [inline test/prep.spec.ts]: Should probably remove console.logs

File: test/init.spec.ts [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    expect(JSON.parse(result.toString())).to.deep.equal(JSON.parse(expected.toString()))
+  })
+
+  it('initalizes existing package.json', () => {
+    const existingPackageJson = path.join(__dirname, 'data', 'existing-package.json')
+    const existingTestPackageJson = path.join(os.tmpdir(), 'existing-package.json')
+    // copy test file to temp dir
+    fs.writeFileSync(existingTestPackageJson, fs.readFileSync(existingPackageJson))
+
+    const publishInfo = {
```
Comment (levthedev) [inline test/init.spec.ts]: Maybe but this in a before so you don't have to declare it twice

**Review by levthedev — CHANGES_REQUESTED** [review]
I approve the PR but I think there are a couple of very small changes that should be made.

File: src/util.ts (hunk 13 above)
Comment (levthedev) [reply src/util.ts]: Especially because identical function above has one

File: .travis.yml (hunk 1 above)
Comment (SimplyAhmazing) [reply .travis.yml]: upgraded to node 8

File: .travis.yml (hunk 2 above)
Comment (SimplyAhmazing) [reply .travis.yml]: removed these

File: appveyor.yml (hunk 4 above)
Comment (SimplyAhmazing) [reply appveyor.yml]: removed

File: src/prep.ts (hunk 10 above)
Comment (SimplyAhmazing) [reply src/prep.ts]: Possible. Made the optimization

File: src/util.ts (hunk 12 above)
Comment (SimplyAhmazing) [reply src/util.ts:19]: So that I can overwrite it for local testing. Also I was thinking it could be potentially useful for supporting other CI's if one knows what env vars are needed.

File: test/init.spec.ts (hunk 18 above)
Comment (SimplyAhmazing) [reply test/init.spec.ts]: I refactored it into  a function. I try not to put input data into `before` but instead use it as a place to setup/teardown services.

File: test/prep.spec.ts (hunk 17 above)
Comment (SimplyAhmazing) [reply test/prep.spec.ts]: removed

File: tsconfig.json (hunk 14 above)
Comment (SimplyAhmazing) [reply tsconfig.json:6]: The declaration setting is used to generate *.d.ts declaration files for your module. However, this setting does not work when you want to use .js files as a part of your code since TS won't know how to infer their module declaration. There's an issue/PR in TS to make this work while allowing for .js code to be apart of the repo (and ignored for type declarations). I think we'll be more likely to use some .js then to need to export types for this module.

File: tsconfig.json (hunk 15 above)
Comment (SimplyAhmazing) [reply tsconfig.json:8]: this allows us to have .js files as a part of module code along with .ts

File: tsconfig.json (hunk 16 above)
Comment (SimplyAhmazing) [reply tsconfig.json:13]: this tells TS to bundle `*.js` files into `dist` dir
